core/src/main/java/org/commonjava/indy/core/change/StoreContentListener.java

```diff
@@ -328,198 +192,109 @@ private int clearPath( String path, ArtifactStore origin, Set<Group> affectedGro
+    private Callable<Integer> clearPathsProcessor( ArtifactStore origin, Predicate<? super String> pathFilter,
+                                          Set<Group> affectedGroups, boolean deleteOriginPath )
     {
-        AtomicInteger count = new AtomicInteger( 0 );
-        try
-        {
-            clearService.drain( partialCount -> count.addAndGet( partialCount ) );
-        }
-        catch ( InterruptedException | ExecutionException e )
-        {
-            logger.error( "Failed to clear paths related to change in " + description, e );
-        }
-
-        logger.debug( "Cleared {} paths for changes in {}", count.get(), description );
-
-        return count.get();
+        return () ->
+            listPathsAnd( origin.getKey(), pathFilter,
+                          p -> clearPath( p, origin, affectedGroups, deleteOriginPath ),
+                          this.directContentAccess );
     }
 
-    /**
-     * Extensive version for clean up paths. This will clean all mergable files in affected groups regardless whether
-     * the path is in the target store cache or not. It also clears NFC.
-     * @param event
-     * @param pathFilter
-     */
-    private void processAllPathsExt( final IndyStoreEvent event, Predicate<? super String> pathFilter )
+    private Predicate<? super String> mergablePath()
     {
-        DrainingExecutorCompletionService<Integer> clearService =
-                new DrainingExecutorCompletionService<>( cleanupExecutor );
-
-        Set<ArtifactStore> stores = new HashSet<>( event.getStores() );
-        stores.forEach(store->{
-            final StoreKey key = store.getKey();
-            try
-            {
-                Set<Group> groups =
-                                storeDataManager.query().packageType( key.getPackageType() ).getGroupsAffectedBy( key );
-                if ( store instanceof Group )
-                {
-                    groups.add( (Group) store );
-                }
-
-                groups.forEach( g->{
-                    listPathsAnd( key, pathFilter, p->clearService.submit(clearPathProcessor( p, key, groups )) );
-                } );
-
-                nfc.clearMissing( LocationUtils.toLocation( store ) ); // clear NFC for this store
-            }
-            catch ( IndyDataException e )
-            {
-                e.printStackTrace();
-            }
-        } );
-
-        Set<StoreKey> keys = event.getStores().stream().map( s -> s.getKey() ).collect( Collectors.toSet() );
-        drainAndCount( clearService, "stores: " + keys );
+        return ( path ) -> {
+            SpecialPathInfo pathInfo = specialPathManager.getSpecialPathInfo( path );
+            return ( pathInfo != null && pathInfo.isMergable() );
+        };
     }
 
-    private Predicate<? super String> mergablePathStrings()
+    /**
+     * Use listable to filter paths of http-metadata, checksum, etc.
```

Comment:
Don't we want to clear these as well?

Reply:
I thought filtering those could improve the perf a bit since we don't cache them in ISPN caches (content-index, nfc, metadata). We should clean them from disk. I will figure a way probably only to filter them when it comes to clean up ISPN.